core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java

```diff
@@ -129,17 +134,23 @@ public RewriteManifests deleteManifest(ManifestFile manifest) {
 
   @Override
   public RewriteManifests addManifest(ManifestFile manifest) {
-    try {
-      // the appended manifest must be rewritten with this update's snapshot ID
-      addedManifests.add(copyManifest(manifest));
-    } catch (IllegalArgumentException e) {
-      throw new IllegalArgumentException("Cannot append manifest: " + e.getMessage());
+    Preconditions.checkArgument(!manifest.hasAddedFiles(), "Cannot add manifest with added files");
+    Preconditions.checkArgument(!manifest.hasDeletedFiles(), "Cannot add manifest with deleted files");
+    Preconditions.checkArgument(manifest.snapshotId() == null, "Snapshot id must be assigned during commit");
```

Comment:
I think this should allow snapshots with ID -1 as well. Previously, we used -1 when the snapshot ID was not set. While it is safe for this to check for null because `ManifestWriter` now produces only snapshots with a valid ID or null, there are situations where manifests were written with a previous version of Iceberg and then are committed with this version. Those manifests would be rejected. Our Flink writer appends manifests that are written like this, so when the application restarts, it may try to append a manifest written in the checkpoint committed prior to the application starting with the new version.

Reply:
@rdblue , do you mean the Flink writer writes snapshot with ID -1 in checkpoint?

Reply:
It used the `ManifestWriter` factory method that created a snapshot with ID -1. Any writer that did this and appended the manifest could get in a bad state by picking up previously staged changes, not just Flink.

Reply:
I think it is reasonable. I updated the condition to allow -1.  Right now, if inheritance is enabled but snapshot ID is -1, we will still rewrite the manifest.

Reply:
I think it is reasonable. I updated the condition to allow -1.  Right now, if inheritance is enabled but snapshot ID is -1, we will still rewrite the manifest.